lib/vquic/ngtcp2.c

```diff
@@ -357,8 +364,16 @@ static int cb_recv_stream_data(ngtcp2_conn *tconn, int64_t stream_id,
     return NGTCP2_ERR_CALLBACK_FAILURE;
   }
 
-  ngtcp2_conn_extend_max_stream_offset(tconn, stream_id, nconsumed);
-  ngtcp2_conn_extend_max_offset(tconn, nconsumed);
+  if(stream_user_data) {
+    struct Curl_easy *data = stream_user_data;
+    struct HTTP *stream = data->req.protop;
+    stream->unacked_window += buflen;
+  }
+  else {
+    /* probably QPACK stream or similar */
+    ngtcp2_conn_extend_max_stream_offset(tconn, stream_id, nconsumed);
```

Comment:
I wonder if we should use `buflen` here for safety, not `nconsumed`.

Reading the docs for [`nghttp3_conn_read_stream`](https://github.com/ngtcp2/nghttp3/blob/5cd9945a0eef23f39fe009cf44dd559438bf2642/lib/includes/nghttp3/nghttp3.h#L1344-L1352) it's not clear under which circumstances, if any, can `nconsumed` be different than `buflen` for a control stream. But to protect against a case where nghttp3 ever returns a value that is different than the size of the input buffer, we should probably use `buflen` since we have been given that many bytes by the QUIC layer and so that's what we should increase the flow control amount by.

Reply:
If we think `buflen` and `nconsumed` must always be the same here, maybe we can add an assert to catch cases where they differ?

Reply:
mr @tatsuhiro-t previously changed *from* buflen to nconsumed in e42bd8a204e4db9320b but yeah, its not superclear what the difference is...

Reply:
nghttp3_conn_read_stream returns the number of bytes inside buflen which consists of framing overhead including QPACK HEADERS.  In other words, it does not consume payload of DATA frame.  It applies to all streams not limited to control streams, so basically ngtcp2_conn_extend_max_stream_offset and ngtcp2_conn_extend_max_offset should be called regardless of stream_user_data.

Reply:
Aha, thanks! Getting the QUIC and HTTP/3 layers untangled and how to deal with flow control windows properly is... complicated!